src/goto-instrument/cover_boundary_values.cpp

```diff
@@ -0,0 +1,314 @@
+    e1.operands().push_back(src.op0());
+    e1.operands().push_back(src.op1());
+    result.insert(e1);
+
+    exprt e2(ID_gt);
+    e2.type().id(src.type().id());
+    e2.operands().push_back(src.op0());
+    e2.operands().push_back(src.op1());
+    result.insert(e2);
+  }
+  // the negation of "<=" is ">"
+  if(src.id()==ID_le)
+  {
+    exprt e1(ID_gt);
+    e1.type().id(src.type().id());
+    e1.operands().push_back(src.op0());
+    e1.operands().push_back(src.op1());
+    result.insert(e1);
+  }
+  // the negation of ">=" is "<"
+  if(src.id()==ID_ge)
+  {
+    exprt e1(ID_lt);
+    e1.type().id(src.type().id());
+    e1.operands().push_back(src.op0());
+    e1.operands().push_back(src.op1());
+    result.insert(e1);
+  }
+  // the negation of ">" is "==" and "<"
+  if(src.id()==ID_gt)
+  {
+    exprt e1(ID_equal);
+    e1.type().id(src.type().id());
+    e1.operands().push_back(src.op0());
+    e1.operands().push_back(src.op1());
+    result.insert(e1);
+
+    exprt e2(ID_lt);
+    e2.type().id(src.type().id());
+    e2.operands().push_back(src.op0());
+    e2.operands().push_back(src.op1());
+    result.insert(e2);
+  }
+  // the negation of "!=" is "=="
+  if(src.id()==ID_notequal)
+  {
+    exprt e1(ID_equal);
+    e1.type().id(src.type().id());
+    e1.operands().push_back(src.op0());
+    e1.operands().push_back(src.op1());
+    result.insert(e1);
+  }
+
+  return result;
+
+}
+
+/*******************************************************************\
+
+Function: non_ordered_expr_expansion
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+std::set<exprt> non_ordered_expr_expansion(const exprt &src)
+{
+  std::set<exprt> result;
+
+  std::set<exprt> s1, s2;
+
+  if(src.id()!=ID_not) s1.insert(src);
+  else
+  {
+    exprt no=src.op0();
+    if(is_arithmetic_predicate(no))
+    {
+      auto res=ordered_negation(no);
+      for(auto &e: res) s1.insert(e);
+    }
+    else result.insert(src);
+  }
+
+  // expand negations and non-ordered predicates
+  while(true)
+  {
+    bool changed=false;
+    for(auto &x: s1)
+    {
+      std::vector<exprt> operands;
+      collect_operands(x, operands);
+      for(size_t i=0; i<operands.size(); i++)
+      {
+        std::set<exprt> res;
+        if(operands[i].id()==ID_not)
+        {
+          exprt no=operands[i].op0();
+          if(is_arithmetic_predicate(no))
+          {
+            res=ordered_negation(no);
+            if(res.size()>0) changed=true;
+          }
+        }
+        else
+        {
+          if(operands[i].id()==ID_le
+             or operands[i].id()==ID_ge
+             or operands[i].id()==ID_notequal)
+          {
+            res=non_ordered_predicate_expansion(operands[i]);
+            if(res.size()>0) changed=true;
+          }
+        }
+        std::set<exprt> co=replacement_and_conjunction(res, operands, i);
+        s2.insert(co.begin(), co.end());
+        if(res.size()>0) break;
```

Comment:
`if(!res.empty()) break;`  (.size() may have linear complexity, empty() is constant time)